skiko/src/jvmMain/cpp/windows/directXRedrawer.cc

```diff
@@ -0,0 +1,343 @@
+        swapChainDesc.SwapEffect = DXGI_SWAP_EFFECT_FLIP_SEQUENTIAL;
+        swapChainDesc.SampleDesc.Count = 1;
+        swapChainDesc.SampleDesc.Quality = 0;
+        swapChainDesc.Scaling = DXGI_SCALING_NONE;
+
+        DXGI_SWAP_CHAIN_FULLSCREEN_DESC swapChainFSDesc = {};
+        swapChainFSDesc.Windowed = TRUE;
+
+        gr_cp<IDXGISwapChain1> swapChain;
+        GR_D3D_CALL_ERRCHECK(swapChainFactory->CreateSwapChainForHwnd(d3dDevice->queue.get(), fWindow, &swapChainDesc, &swapChainFSDesc, nullptr, &swapChain));
+        GR_D3D_CALL_ERRCHECK(swapChain->QueryInterface(IID_PPV_ARGS(&d3dDevice->swapChain)));
+        GR_D3D_CALL_ERRCHECK(d3dDevice->device->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&d3dDevice->fence)));
+        d3dDevice->fenceEvent = CreateEvent(nullptr, FALSE, FALSE, nullptr);
+
+        return toJavaPointer(d3dDevice);
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_resizeBuffers(
+        JNIEnv *env, jobject redrawer, jlong devicePtr, jint width, jint height)
+    {
+        DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice*>(devicePtr);
+        if (!d3dDevice->isSizeEqualTo(width, height))
+        {
+            GR_D3D_CALL_ERRCHECK(d3dDevice->swapChain->ResizeBuffers(BuffersCount, width, height, DXGI_FORMAT_R8G8B8A8_UNORM, 0));
+            d3dDevice->bufferWidth = width;
+            d3dDevice->bufferHeight = height;
+        }
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_finishFrame(
+        JNIEnv *env, jobject redrawer, jlong devicePtr, jlong contextPtr, jlong surfacePtr)
+    {
+        DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice*>(devicePtr);
+
+        SkSurface *surface = fromJavaPointer<SkSurface*>(surfacePtr);
+        surface->flushAndSubmit();
+        GrDirectContext *fContext = fromJavaPointer<GrDirectContext*>(contextPtr);
+        surface->flush(SkSurface::BackendSurfaceAccess::kPresent, GrFlushInfo());
+        fContext->flush({});
+        fContext->submit(true);
+
+        // 1 value in [Present(1, 0)] enables vblank wait so this is how vertical sync works in DirectX.
+        GR_D3D_CALL_ERRCHECK(d3dDevice->swapChain->Present(1, 0));
+
+        const UINT64 fence = d3dDevice->fenceValue;
+        GR_D3D_CALL_ERRCHECK(d3dDevice->queue->Signal(d3dDevice->fence.get(), fence));
+        d3dDevice->fenceValue++;
+        if (d3dDevice->fence->GetCompletedValue() < fence)
+        {
+            GR_D3D_CALL_ERRCHECK(d3dDevice->fence->SetEventOnCompletion(fence, d3dDevice->fenceEvent));
+            WaitForSingleObjectEx(d3dDevice->fenceEvent, INFINITE, FALSE);
+        }
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_disposeDevice(
+        JNIEnv *env, jobject redrawer, jlong devicePtr)
+    {
+        DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice*>(devicePtr);
+        delete d3dDevice;
```

Comment:
d3dDevice can be 0, so we need to check that

Reply:
This is not necessary here. If `d3dDevice` is `NULL`, the `delete` operator does nothing.

Reply:
Didn't knpw that :)

https://stackoverflow.com/questions/6731331/is-it-still-safe-to-delete-nullptr-in-c0x